File: src/bin/c29_exercise.rs
pub fn longer<'a>(a: &'a str, b: &'a str) -> &'a str {
    // TODO: Return whichever of `a` or `b` has the longer length.
    // If they're equal, return `a`.
    // The lifetime 'a means: the output reference lives at least as long
    // as the shorter-lived of the two inputs.
    if a.len() >= b.len() {a} else {b}
}

fn main() {
    let s1 = "hello";
    let s2 = "hi";
    println!("{}", longer(s1, s2));
}

File: python_exercises/c29_exercise.py
# Rust Concept: Lifetimes ('a)
#
# N/A in Python. Rust lifetimes tell the compiler how long references are valid:
#   fn longer<'a>(a: &'a str, b: &'a str) -> &'a str
#
# This means: "the returned reference lives as long as the shorter-lived input."
# The compiler uses this to prevent dangling references at compile time.
#
# Python has garbage collection — objects live as long as something references them.
# No dangling pointers, no lifetime annotations, no compile-time checking.
# The tradeoff: Python pays for GC overhead at runtime. Rust pays zero.
